c++/src/kj/mutex.c++

```diff
@@ -164,12 +177,20 @@ void Mutex::lock(Exclusivity exclusivity) {
 
         // The mutex is exclusively locked by another thread.  Since we incremented the counter
         // already, we just have to wait for it to be unlocked.
-        syscall(SYS_futex, &futex, FUTEX_WAIT_PRIVATE, state, nullptr, nullptr, 0);
+        auto result = syscall(SYS_futex, &futex, FUTEX_WAIT_PRIVATE, state, specp, nullptr, 0);
+        if (result < 0) {
+          // If we timeout though, we need to signal that we're not waiting anymore.
+          if (errno == ETIMEDOUT) {
+            state = __atomic_sub_fetch(&futex, 1, __ATOMIC_RELAXED);
```

Comment:
Hmmmmmm... what if, between timing out, and calling this to decrement the reader count, the mutex was unlocked, and also this decrement sets the reader count to 0, and also another thread in the meantime was waiting for an exclusive lock?

I think we need to perform the same sequence as in unlock() after the `__atomic_sub_fetch`, to make sure we send a wake signal if needed.